File: TransCoder_tokenized_test_set_functions/python/GIVEN_LARGE_NUMBER_CHECK_SUBSEQUENCE_DIGITS_DIVISIBLE_8_1.py
def isSubSeqDivisible ( str ) :
    n = len ( str )
    dp = [ [ 0 for i in range ( 10 ) ] for i in range ( n + 1 ) ]
    arr = [ 0 for i in range ( n + 1 ) ]
    for i in range ( 1, n + 1 ) :
        arr [ i ] = int ( str [ i - 1 ] );
        
    for i in range ( 1, n + 1 ) :
        dp [ i ] [ arr [ i ] % 8 ] = 1;
        
        for j in range ( 8 ) :
            if ( dp [ i - 1 ] [ j ] > dp [ i ] [ ( j * 10 + arr [ i ] ) % 8 ] ) :
                dp [ i ] [ ( j * 10 + arr [ i ] ) % 8 ] = dp [ i - 1 ] [ j ]
            if ( dp [ i - 1 ] [ j ] > dp [ i ] [ j ] ) :
                dp [ i ] [ j ] = dp [ i - 1 ] [ j ]
    for i in range ( 1, n + 1 ) :
        if ( dp [ i ] [ 0 ] == 1 ) :
            return True
    return False

File: TransCoder_tokenized_test_set_functions/java/GIVEN_LARGE_NUMBER_CHECK_SUBSEQUENCE_DIGITS_DIVISIBLE_8_1.java
static boolean isSubSeqDivisible ( String str ) {
  int n = str . length ( );
  int dp [ ] [ ] = new int [ n + 1 ] [ 10 ];
  int arr [ ] = new int [ n + 1 ];
  for ( int i = 1;
  i <= n;
  i ++ ) arr [ i ] = ( int ) ( str . charAt ( i - 1 ) - '0' );
  for ( int i = 1;
  i <= n;
  i ++ ) {
    dp [ i ] [ arr [ i ] % 8 ] = 1;
    for ( int j = 0;
    j < 8;
    j ++ ) {
      if ( dp [ i - 1 ] [ j ] > dp [ i ] [ ( j * 10 + arr [ i ] ) % 8 ] ) dp [ i ] [ ( j * 10 + arr [ i ] ) % 8 ] = dp [ i - 1 ] [ j ];
      if ( dp [ i - 1 ] [ j ] > dp [ i ] [ j ] ) dp [ i ] [ j ] = dp [ i - 1 ] [ j ];
    }
  }
  for ( int i = 1;
  i <= n;
  i ++ ) {
    if ( dp [ i ] [ 0 ] == 1 ) return true;
  }
  return false;
}